command/agent/dns.go

```diff
@@ -664,6 +665,9 @@ func (d *DNSServer) serviceNodeRecords(dc string, nodes structs.CheckServiceNode
 	qName := req.Question[0].Name
 	qType := req.Question[0].Qtype
 	handled := make(map[string]struct{})
+
+	// Post-shuffle: limit the number of nodes we return to the client
+	effectiveRecordLimit := lib.MinInt(d.config.UDPAnswerLimit, maxUDPAnswerLimit)
```

Comment:
This will affect the length of TCP responses, too - looking at this I think you'd need to pull all the logic here up one level (basically have the truncation logic a-la https://github.com/hashicorp/consul/pull/1712/files#diff-08f6dd8ba6ebd703aba0bfcbf0f2abd4R646 respect the new config parameter).